src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,276 @@
+            finally
+            {
+                stack.Dispose();
+            }
+
+            return new JsonDocument(utf8Json, database, extraRentedBytes);
+        }
+
+        private static ArraySegment<byte> ReadToEnd(Stream stream)
+        {
+            int written = 0;
+            byte[] rented = null;
+            ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+
+            try
+            {
+                long expectedLength = 0;
+
+                if (stream.CanSeek)
+                {
+                    expectedLength = Math.Max(1L, stream.Length - stream.Position);
+                    rented = pool.Rent(checked((int)expectedLength));
+                }
+                else
+                {
+                    rented = pool.Rent(UnseekableStreamInitialRentSize);
+                }
+
+                int lastRead;
+
+                do
+                {
+                    if (expectedLength == 0 && rented.Length == written)
+                    {
+                        byte[] tmp = pool.Rent(rented.Length * 2);
+                        Buffer.BlockCopy(rented, 0, tmp, 0, rented.Length);
+                        rented.AsSpan().Clear();
+                        pool.Return(rented);
+                        rented = tmp;
+                    }
+
+                    lastRead = stream.Read(rented, written, rented.Length - written);
+                    written += lastRead;
+                } while (lastRead > 0);
+
+                return new ArraySegment<byte>(rented, 0, written);
+            }
+            catch
+            {
+                if (rented != null)
+                {
+                    rented.AsSpan(0, written).Clear();
+                    pool.Return(rented);
+                }
+
+                throw;
+            }
+        }
+
+        private static async Task<ArraySegment<byte>> ReadToEndAsync(
+            Stream stream,
+            CancellationToken cancellationToken)
+        {
+            int written = 0;
+            byte[] rented = null;
+            ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+
+            try
+            {
+                long expectedLength = 0;
+
+                if (stream.CanSeek)
+                {
+                    expectedLength = Math.Max(1L, stream.Length - stream.Position);
+                    rented = pool.Rent(checked((int)expectedLength));
+                }
+                else
+                {
+                    rented = pool.Rent(UnseekableStreamInitialRentSize);
+                }
+
+                int lastRead;
+
+                do
+                {
+                    if (expectedLength == 0 && rented.Length == written)
+                    {
+                        byte[] tmp = pool.Rent(rented.Length * 2);
+                        Buffer.BlockCopy(rented, 0, tmp, 0, rented.Length);
+                        rented.AsSpan().Clear();
+                        pool.Return(rented);
+                        rented = tmp;
```

Comment:
It shouldn't matter on .NET Core where thread aborts aren't supported, but in case this code ends up getting used elsewhere, the paranoid side of me worries that an exception above could cause pool.Return(rented) to return the buffer, and then the catch block to return the buffer again.  I suggest instead:
```C#
byte[] toReturn = rented;
rented = ArrayPool<byte>.Shared.Rent(toReturn.Length * 2);
Buffer.BlockCopy(toReturn, 0, rented, 0, toReturn.Length);
ArrayPool<byte>.Shared.Return(toReturn, clearArray: true);
```